library/src/test/java/io/voucherify/android/client/VoucherifyUtilsTest.java

```diff
@@ -0,0 +1,291 @@
+    }
+
+    @Test
+    public void calculateDiscountAmount() throws Exception {
+        testDiscountAmount("discount < price", 10.0, 10, 0.1);
+        testDiscountAmount("discount == price", 10.0, 1000, 10.0);
+        testDiscountAmount("discount > price", 10.0, 2000, 10.0);
+        testDiscountAmount("discount == 0", 10.0, 0, 0.0);
+    }
+
+    @Test
+    public void calculateDiscountUnit() throws Exception {
+        testDiscountUnit("unitOff > 0", 10.0, 1.0, 2.0, 2.0);
+        testDiscountUnit("unitOff == 0", 10.0, 1.0, 0.0, 0.0);
+    }
+
+    @Test
+    public void calculateDiscountGift() throws Exception {
+        testDiscountGift("giftBalance < price", 10.0, 100, 1.0);
+        testDiscountGift("giftBalance == 0", 10.0, 0, 0.0);
+        testDiscountGift("giftBalance == price", 10.0, 10000, 10.0);
+        testDiscountGift("giftBalance > price", 10.0, 20000, 10.0);
+        testDiscountGift("giftBalance < 0", 10.0, -100, -1.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountInvalidType() throws Exception {
+        BigDecimal basePrice = new BigDecimal(10);
+        Discount discount = createDiscount(null, null, null, null, null);
+        VoucherResponse voucherResponse = createVoucherResponse("", true, discount, null, null, null);
+        VoucherifyUtils.calculateDiscount(basePrice, voucherResponse, null);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountPercentageMoreThanHundred() throws Exception {
+        testDiscountPercentage("discount > 100", 10.0, 110.0, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountPercentageLessThanZero() throws Exception {
+        testPricePercentage("discount < 0", 10.0, -10.0, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountPercentageNull() throws Exception {
+        testDiscountPercentage("discount == null", 10.0, null, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountAmountLessThanZero() throws Exception {
+        testDiscountAmount("discount < 0", 10.0, -10, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountAmountNull() throws Exception {
+        testDiscountAmount("discount == null", 10.0, null, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountUnitLessThanZero() throws Exception {
+        testDiscountUnit("unitOff < 0", 10.0, 1.0, -2.0, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountUnitNull() throws Exception {
+        testDiscountUnit("unitOff == null", 10.0, 1.0, null, 0.0);
+    }
+
+    @Test(expected = RuntimeException.class)
+    public void calculateDiscountGiftNull() throws Exception {
+        testDiscountGift("giftBalance == null", 10.0, null, 0.0);
+    }
+
+    private void testPricePercentage(String description,
```

Comment:
In my opinion, we should move the assert to the testing method as it is more readable. This and the below methods should be rather extracted to a test util class which helps to prepare some data during testing.